### PR TITLE
File attachment

### PR DESCRIPTION
**Pull request by lasryaric**
# Add support for file attachment as content fragment in conversations.

_(Does not build in CI because I modified Sparkle directly in this branch, I will fix that ASAP, but you can still review the code. It builds otherwise)._

- Support for PDF and plain/text files only as per [framing](https://github.com/dust-tt/dust/issues/2548).
- Add a file picker in the UI.
- Adding full context (username, timezone, profilerURL etc) to content fragment.
- [Video demo.](https://www.loom.com/share/ae25da820de54b92ae37c6a9af9aef3d)
- Screenshot:

Screen Shot 2023-11-15 at 23 09 48

# Deployment
- Deploy Sparkle.
- Update front/package.json @dust-tt/sparkle dependency.
- `front> npm run initdb`
- Deploy front.
- Deploy connectors.

### CODE REVIEW

File: connectors/src/connectors/slack/bot.ts
Comment (fontanierh) [inline connectors/src/connectors/slack/bot.ts:723]: Why the "as" here ? can't we have type-checking ?

File: front/components/assistant/conversation/InputBar.tsx
Comment (fontanierh) [inline front/components/assistant/conversation/InputBar.tsx:281]: can we remove the any here ?

File: front/components/assistant/conversation/InputBar.tsx
Comment (fontanierh) [inline front/components/assistant/conversation/InputBar.tsx:285]: can we remove the any here ?

File: front/components/assistant/conversation/InputBar.tsx
Comment (fontanierh) [inline front/components/assistant/conversation/InputBar.tsx:294]: can we remove the any here ?

File: front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts
Comment (fontanierh) [inline front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts:15]: why are we duplicating this ?

**Review by fontanierh — CHANGES_REQUESTED** [review]
Looks great overall ! Just a few comments

File: connectors/src/connectors/slack/bot.ts
Comment (lasryaric) [reply connectors/src/connectors/slack/bot.ts:723]: @fontanierh I don't see why this removes type checking.
This is similar to typing the function explicitly. If something is not compliant with the type `PostContentFragmentRequestBody`, it will complain.

File: front/components/assistant/conversation/InputBar.tsx
Comment (lasryaric) [reply front/components/assistant/conversation/InputBar.tsx]: Sorry, copy/pasted.

File: connectors/src/connectors/slack/bot.ts
Comment (fontanierh) [reply connectors/src/connectors/slack/bot.ts:723]: No it is not. `as X` means "STFU typescript, this is what I say it is, don't check it"

Comment (lasryaric) [reply connectors/src/connectors/slack/bot.ts]: I suppose it depends on the situation, but if typescript can do type checking it will do it apparently.
If I have a typo in my payload, it will complain.
But I understand this might not be the most expressive way to do that, and might lead to error.


Screen Shot 2023-11-16 at 11 14 51

Comment (lasryaric) [reply connectors/src/connectors/slack/bot.ts]: Fixed.

File: front/components/assistant/conversation/InputBar.tsx
Comment (lasryaric) [reply front/components/assistant/conversation/InputBar.tsx]: Sorry, copy/pasted.

File: front/components/assistant/conversation/InputBar.tsx
Comment (lasryaric) [reply front/components/assistant/conversation/InputBar.tsx]: Sorry, copy/pasted.

File: front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts
Comment (lasryaric) [reply front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts:15]: @fontanierh between api/v1 and api/ these are not exactly the same.
In v1, we have properties that we derive from the user (`PostContentFragmentRequestBodySchema.context.email` for example) while in the api/v1 case, we don't have a user object and we have to rely on the user input.

**Review by lasryaric — COMMENTED** [review]
addressed your comments @fontanierh

File: connectors/src/connectors/slack/bot.ts
Comment (fontanierh) [reply connectors/src/connectors/slack/bot.ts]: Interesting, thanks ! Didn't know that it "sometimes" checks 🤔  Not sure I fully understand how this works

File: front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts
Comment (fontanierh) [reply front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts:15]: Got it 👍

**Review by fontanierh — DISMISSED** [review]
LGTM modulo lint errors 🙏

File: connectors/src/connectors/slack/bot.ts
Comment (lasryaric) [reply connectors/src/connectors/slack/bot.ts]: Same. I would like to understand what are the rules here. But I also agree that this can be miss leading so I can stop doing that.

Comment (fontanierh) [reply connectors/src/connectors/slack/bot.ts]: I've definitely seen a bunch of cases where it allows to lie about the type. 
For instance, this:
Screenshot 2023-11-16 at 11 49 23

**Review by philipperolet — CHANGES_REQUESTED** [review]
Big work for a great feature :pray: Still reviewing but giving you here first issues so you can address
- "File not supported" when I try a .md file but the notifs mentions .md as supported
- Tried with 2 pdfs, got "Failed to import PDF" (they seemed pretty basic)
- After a failure if I close the failure notification, retrying seems broken (no attachment nor notification of failure)
Available to discuss at your convenience

**Comment by lasryaric** [comment]
@philipperolet I did a few changes to make typescript happy about the `any` suggested by @fontanierh and did not re-test everything, so I might have added a little faulty check. 
I have a[ test suit](https://github.com/orgs/dust-tt/projects/3?pane=issue&itemId=44717993) to run before merging anyway, so it would not have gone to production but thanks for reporting them, I will fix it asap.

**Comment by philipperolet** [comment]
> @philipperolet I did a few changes to make typescript happy about the `any` suggested by @fontanierh and did not re-test everything, so I might have added a little faulty check. I have a[ test suit](https://github.com/orgs/dust-tt/projects/3?pane=issue&itemId=44717993) to run before merging anyway, so it would not have gone to production but thanks for reporting them, I will fix it asap.

**Comment by philipperolet** [comment]
No close button on this one
image

**Comment by philipperolet** [comment]
Also seems to block down scroll for all message
image

**Comment by lasryaric** [comment]
> No close button on this one


@philipperolet did you npm install? I updated sparkle

**Comment by philipperolet** [comment]
I did. Will redo

**Comment by philipperolet** [comment]
**Product considerations**
I understand we support only one file. Then I would disable the attachment button when the file has already been selected (and reenable on removal) => with tooltip at the time only one file is supported

As a user, feels weird when I want to update a second doc that it replaces the first one that I had not given up

Aslo, would add a tooltip for the attachment button to clarify what it does to noobs: "Add a file (.txt, .pdf or .md) to your message that the assistants will use for answering" or something like that

File: front/pages/api/v1/w/[wId]/assistant/conversations/[cId]/content_fragments.ts
Comment (spolu) [inline front/pages/api/v1/w/[wId]/assistant/conversations/[cId]/content_fragments.ts:27]: should be optional

File: front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts
Comment (spolu) [inline front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts:24]: why take timezone ?
why take profilePictureUrl only ?

From slack we want 0 context. It's not user related.

For user provided contentFragment we want user information which I think should cover as much as possible (userame display name so that we can render a hover if needed).

Timezone is not needed here so I would overall push here profile picture username and name but not timeZone. Not push anything from slack. And in DB remove timeZone

File: front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts
Comment (spolu) [inline front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts:33]: What's this?

File: front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts
Comment (spolu) [inline front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts:75]: You can merge the 2 errors in one condition

File: front/pages/api/w/[wId]/assistant/conversations/index.ts
Comment (spolu) [inline front/pages/api/w/[wId]/assistant/conversations/index.ts:189]: Within the product we don't need an updated view, it will do so on its own. So we can save that get

File: front/pages/api/w/[wId]/assistant/conversations/index.ts
Comment (spolu) [inline front/pages/api/w/[wId]/assistant/conversations/index.ts:155]: I don't think we need that one either

File: sparkle/src/components/Citation.tsx
Comment (spolu) [inline sparkle/src/components/Citation.tsx:1]: this should be all cleaned-up

File: front/pages/w/[wId]/assistant/new.tsx
Comment (spolu) [inline front/pages/w/[wId]/assistant/new.tsx:109]: what is `url` here?

**Review by spolu — CHANGES_REQUESTED** [review]
Where is the code that uploads the file client side?

File: front/components/assistant/conversation/InputBar.tsx
Comment (spolu) [inline front/components/assistant/conversation/InputBar.tsx:281]: any chance we can dry this?

**Review by spolu — COMMENTED** [review]
InputBar looks reasonable would be nice to dry the upload logic to avoid cluttering this already massive file :pray:

File: sparkle/src/components/Citation.tsx
Comment (philipperolet) [inline sparkle/src/components/Citation.tsx]: What is the purpose of this line?

File: front/components/assistant/conversation/InputBar.tsx
Comment (philipperolet) [inline front/components/assistant/conversation/InputBar.tsx]:
```suggestion
           <Tooltip label="Add a file (.txt, .pdf or .md) to your message that assistants can use" position="above">
              <IconButton
                className="element-700 block"
                variant={"tertiary"}
                icon={AttachmentStrokeIcon}
                size="md"
                disabled={false}
                onClick={() => {
                  fileInputRef.current?.click();
                }}
              />
            </Tooltip>
```

Comment (philipperolet) [reply front/components/assistant/conversation/InputBar.tsx]: Here's a working tooltip :) The keyword du jour was "block"

**Review by philipperolet — DISMISSED** [review]
LGTM modulo @spolu 's comments & a question
Also :gift: added a working tooltip :slightly_smiling_face:  (1-click commit but for the import of Tooltip)

File: front/components/assistant/conversation/InputBar.tsx
Comment (lasryaric) [reply front/components/assistant/conversation/InputBar.tsx]: dried!

File: front/components/assistant/conversation/InputBar.tsx
Comment (philipperolet) [reply front/components/assistant/conversation/InputBar.tsx]: I also removed the "s-" in className that is only for sparkle

File: front/pages/w/[wId]/assistant/new.tsx
Comment (lasryaric) [reply front/pages/w/[wId]/assistant/new.tsx]: removed

File: sparkle/src/components/Citation.tsx
Comment (lasryaric) [reply sparkle/src/components/Citation.tsx]: cleaned up

File: front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts
Comment (lasryaric) [reply front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts:24]: Cleaned up

File: front/components/assistant/conversation/InputBar.tsx
Comment (philipperolet) [reply front/components/assistant/conversation/InputBar.tsx]: :warning: the suggestion does not contain the tooltip import though

**Comment by lasryaric** [comment]
thank you very much for the Tooltip and the review @philipperolet .

@spolu, I addressed all your comments.

**Review by lasryaric — COMMENTED** [review]
(Posting my answers)

**Review by spolu — DISMISSED** [review]
LGTM

Can you check @philipperolet is A-OK also :pray:

**Comment by lasryaric** [comment]
@philipperolet are you ok?

File: front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts
Comment (philipperolet) [reply front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts:24]: It seems there's still timezone & not other fields? maybe I missed something you discussed irl?

File: front/pages/api/w/[wId]/assistant/conversations/index.ts
Comment (philipperolet) [inline front/pages/api/w/[wId]/assistant/conversations/index.ts:24]: nit: relative import

**Review by philipperolet — CHANGES_REQUESTED** [review]
LGTM :+1: Marked request changes just because there's a com from stan on which I don't understand how it was adressed (cf below) => feel free to discard if I missed something, no need for ptal (other comment is a nit)

File: front/components/assistant/conversation/InputBar.tsx
Comment (philipperolet) [inline front/components/assistant/conversation/InputBar.tsx:454]: just saw you put the 's-' back? I had removed the 's-' and explained in the comments the 's-' is in sparkle/ not elsewhere. Do you disagree?

Comment (philipperolet) [reply front/components/assistant/conversation/InputBar.tsx:454]: Oh ok, it's to override the s-element-600 imposed by sparkle. IMHO a bit shaky, I would remove the class altogether and let it be s-text-element-600 from sparkle, or if important make the fix to be able to do text-element properly -- but feel free to adress how you see fit

=> however it's `text-element` not just `element` (might be a mistake from my commit suggestion if yes apologies) so this has to be changed since just 'element-700' / 's-element-700' are not real classes

Comment (philipperolet) [reply front/components/assistant/conversation/InputBar.tsx:454]: Not sure i'm being clear :sweat_smile: happy to explain first thing tom morning if needed

Comment (spolu) [reply front/components/assistant/conversation/InputBar.tsx:454]: Yep we should have no s- in front

Comment (spolu) [reply front/components/assistant/conversation/InputBar.tsx:454]: This should be part of sparkle as an option if needed

Comment (lasryaric) [reply front/components/assistant/conversation/InputBar.tsx:454]: I am removing the `s-`. I did not disagree, I just did not copy/paste your code @philipperolet.

File: front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts
Comment (lasryaric) [reply front/pages/api/w/[wId]/assistant/conversations/[cId]/content_fragment/index.ts:24]: Removed. thanks.

**Comment by lasryaric** [comment]
Merging. Thank you everybody (@philipperolet @spolu @fontanierh @Duncid @gabhubert )

**Comment by gabhubert** [comment]
ooooooh

**Comment by gabhubert** [comment]
image

**Pull request merged**
